### DISCUSSION TITLE
Defining individual constraints for each row of a decision matrix variable and general constraints for the whole matrix in one optimization problem

### DISCUSSION THREAD

**Discussion by nbahari** [yalmip/YALMIP#1097] [Problems using YALMIP] [answered]
Edit: I tried to delete this discussion, because I noticed this question is related to category of Optimization Modeling, not here. Sorry for the inconvenience!

Hello All, 

Suppose we have a decision matrix variable **P = sdpvar(m,n)**. We have the constraints **c_1, c_2,...,c_m** each of which being applied on the respective row of matrix **P**. In addition, we have a set of constraints **B** which are being applied on the whole matrix. 
For example, can we have such a way of defining constraints?
If yes, is it efficient in large scale and more complex problems? And also, does it yield accurate solutions?
```
n = 3;
m= 2;
P =  sdpvar(m,n);

m_1_max =  2;
m_1_min = -2;

m_2_max = 5;
m_2_min = -5;

P_max = 3;

% Defining the constraints for the first row

 m_1_min <= P(1,:) <= m_1_max;
 
 % Defining the constraints for the second row
 
 m_2_min <= P(2,:) <= m_2_max;
 
 % Defining the constraints for P
 
 P(:) <= P_max;
```

**Comment by johanlofberg** [comment, marked as answer]
Sure, that is a sensible approach, although you of course can vectorize it if you want

Whether the solver computes a good solution or not is impossible to answer, as always.

**Reply by nbahari** [reply]
Thank you Mr. @johanlofberg !! If it yields an accuare solution, it will be a very efficient approach in my project!

**Reply by nbahari** [reply]
@johanlofberg . I would be thankful if you could help me with this! 
I have:
```
nu = 4;
L= 10;
ID = 1;
U = sdpvar(nu, 2*L-1,'full');
```
I define a constraint for U(:,k) as follows:
`Fu = [ Fu, Sys.u_lb(iu) <= U(iu,:) <= Sys.u_ub(iu)] `

The line above works just fine but I get an error for the following line which I do not get why.

`Fs =[Fs, U(ID,k) == 2];`

The error is:

```
Error using sdpvar/subsref (line 161)
Error using sdpvar/subsref>subsref2d (line 306)
Index exceeds matrix dimensions.
```
I do not see any exceeding in the dimesnion of matrix U.

**Reply by johanlofberg** [reply]
You would have to supply a reproducible example, as this does not run (code looks standard, hence there is something fishy with ID or k and the dimension of U)

**Reply by nbahari** [reply]
The whole code is very huge to be put here. I should look for bugs in other places, I think. Thank you for replying!

**Reply by johanlofberg** [reply]
All you have to do is to show/look at ID and k before that line which crashes. No need for all the code as the error must be related to ID and k

**Reply by nbahari** [reply]
Yes, you are right! k reaches the last index of U before the loop ends. It was simple thing to check. Sorry for the inconvenience!
